oak_restricted_kernel/src/syscall/dice_data.rs

```diff
@@ -16,42 +16,123 @@
+        match self {
+            #[cfg(feature = "initrd")]
+            DiceDataDescriptor::Writeable(_write_state) => Err(Errno::EINVAL),
+            DiceDataDescriptor::Readable(read_state) => {
+                let data_as_slice = read_state.data.as_mut_slice();
+                let length = copy_max_slice(&data_as_slice[read_state.index..], buf);
 
-        // destroy the data that was read, to ensure that it can only be read once
-        let slice_to_read = &mut data_as_slice[self.index..(self.index + length)];
-        slice_to_read.zeroize();
+                // destroy the data that was read, to ensure that it can only be read once
+                let slice_to_read =
+                    &mut data_as_slice[read_state.index..(read_state.index + length)];
+                slice_to_read.zeroize();
 
-        self.index += length;
-        Ok(length as isize)
+                read_state.index += length;
+                Ok(length as isize)
+            }
+        }
     }
 
-    fn write(&mut self, _buf: &[u8]) -> Result<isize, oak_restricted_kernel_interface::Errno> {
-        // Writing is not supported.
-        Err(Errno::EINVAL)
+    fn write(&mut self, buf: &[u8]) -> Result<isize, oak_restricted_kernel_interface::Errno> {
+        match self {
+            DiceDataDescriptor::Readable(read_state) => match &mut read_state.data {
+                #[cfg(feature = "initrd")]
+                DiceData::Layer0(stage0_dice_data) => {
+                    <Stage0DiceData as zerocopy::AsBytes>::as_bytes_mut(stage0_dice_data).zeroize();
+                    let _ = core::mem::replace(
+                        self,
+                        Self::Writeable(Box::new(WriteState {
+                            index: 0,
+                            data: <RestrictedKernelDiceData as zerocopy::FromZeroes>::new_zeroed(),
+                        })),
+                    );
+                    self.write(buf)
+                }
+                _ => Err(Errno::EINVAL),
+            },
+            #[cfg(feature = "initrd")]
+            DiceDataDescriptor::Writeable(write_state) => {
+                let data_as_slice = <RestrictedKernelDiceData as zerocopy::AsBytes>::as_bytes_mut(
+                    &mut write_state.data,
+                );
+                let length = copy_max_slice(buf, &mut data_as_slice[write_state.index..]);
+
+                write_state.index += length;
+
+                if write_state.index > data_as_slice.len() {
+                    // the dice data the app may write has a known size. If the app keeps writing to
+                    // this FD after it has written all of it, it's doing something wrong.
+                    return Err(Errno::EINVAL);
+                }
+
+                if write_state.index == data_as_slice.len() {
+                    let read_data =
+                        <RestrictedKernelDiceData as zerocopy::FromBytes>::read_from(data_as_slice)
+                            .unwrap();
```

Comment:
This is an interesting situation, one where I think the unsafe implementations behind `zerocopy` leak from the safe abstraction. That or perhaps it creates an internal copy (in defiance of its name).

Instead of creating a new struct, I could pass the existing instance of `RestrictedKernelDiceData` on the following lines (120-126):

```
                    let _ = core::mem::replace(
                        self,
                        Self::Readable(Box::new(ReadState {
                            index: 0,
                            data: DiceData::Layer1(Box::new(write_state.data)),
                        })),
                    );
```

which of course yields the following error:

```
--> /usr/local/google/home/julsh/oak/oak_restricted_kernel/src/syscall/dice_data.rs:121:25
    |
105 |             DiceDataDescriptor::Writeable(write_state) => {
    |                                           ----------- first mutable borrow occurs here
...
121 |                         self,
    |                         ^^^^ second mutable borrow occurs here
...
124 |                             data: DiceData::Layer1(Box::new(write_state.data)),
    |                                                             ---------------- first borrow later used here

error[E0507]: cannot move out of `write_state.data` which is behind a mutable reference
   --> /usr/local/google/home/julsh/oak/oak_restricted_kernel/src/syscall/dice_data.rs:124:61
    |
124 | ...                   data: DiceData::Layer1(Box::new(write_state.data)),
    |                                                       ^^^^^^^^^^^^^^^^ move occurs because `write_state.data` has type `RestrictedKernelDiceData`, which does not implement the `Copy` trait

Some errors have detailed explanations: E0499, E0507.
For more information about an error, try `rustc --explain E0499`.
error: could not compile `oak_restricted_kernel` (lib) due to 2 previous errors
```

Creating a new struct from the slice of `write_state.data` silences this complaint from the compiler.